### DISCUSSION TITLE
对组件 options 类型的 2 种设计

### DISCUSSION THREAD

**Discussion by timongh** [the1812/Bilibili-Evolved#3011] [分享 / Share]
我设计了两套组件 options 的类型系统。两套类型都可以让 entry 的参数获得类型提示。

不看原理，只看使用方式也好，都在代码里面。看看大家更喜欢哪一种。如果有更好的方案也欢迎提出哦。

- 方案 1：这套方案特点是直接，容易理解。[代码](https://www.typescriptlang.org/zh/play?#code/LYewJgrgNgpgBAYxMADiAdjdAXAjHAbwCg5S4BLVEAJ2zgBE4BeRZNTHAYWpgENsYREmQD0IuICztQJJygEb9Ab3LDSMAB5pardAGc6SKhzot6AOgQ9+MTmwxZsAWRjZeYfrwAUxMp7jpewGAC44AHJddhtcADlfGCCAGkUvEBRscgxNQI8vLNDrHFx6FwyErKywGAAzXmhsADVeKAgAuAAGeJKvAF82kq7i0XEbagBPOEAQt0Ah5UAHU0B75UBTuUB24MB0-UBvH0Bo9UAF40AuTz7SMThAU+i1wFNzQCvAuCSUtMkpQFo5Ubgyyuq6hsF2waHA1xAASmYAPnO3VIHR+RA6QlAkFgrD0NgATIQEpQ1HRGCwcvpuHwBEJPHtABKK0nOyVSWgSSC0dAuZM0zAYJjMAgA8qS0vZHM5HO4drCwjh4QVHEV2p5HlUoLV6o1Aq1eb1PKC8f04CdABVKgArjQC1pnBWZctGM1glsEMUPBLHCBXrafTjNa0szyhzeAAeE1mkDlEn6zT-ZW7AY4YZwQAm1oA302mgAp1QBomoALhOkgBh-wAocoBIczWWwpaToH3p30CFv52Hh9q0fyYgJA-rgKlRGm0fNyBgZphxFis+mdXLcmSyPj8gRC7YRUT8cV5NLSwK8HynSohRCh0HgmJs2PMSM8KJodAAKvSVzgd6bBAka9u4OUIOgEPrEEyYCXNJ2XC7mdXlAJ0GA6TujABVdAAGt0BAAB3dBH3+VxeQAN0CX9H2fRxXz9TwfngoxEIcJwX2ZQEey8HhsAgah0DgGCEnnTwz3US9r1vFtzALRskNdN8VE-b84F-ADgLAiC2S0KDYIw5iO2wrsUISdDuKMMSbFYlCNxKIiSLIijFSrGi6HdeBH0dZ0XVsd9OJ-TDBKfCSX14dAhn+QEWGg9pjI4rAuIQizFPIdByhgahdUBAB+XU4ECTAYL86TwUhcAlwbfQjzNZSlFUc9vIEahKgQeBeJA8DH2SrwAG0+yabRqG8gBzABdQIrz48DKK01L1HSvysr0wTFJqel6ry9B8N5cVnmlJoaia08Wp049dU8qzkPYj83LpXL+IKlhVvyiyHMKzwipHeBvLgQCYCGT1dVq2b9UU5l9uiarULIKiyG0igcHa3hsrgeScEU3lFtMuBNoEn16WBx8EkG9pSsCcqqvHCzAiwzlcMerIPi+CctCRstARgkByDACbqKmt6Mo676hxwABlBwUnQSrNBdf6TOWoGgP69b2YakHaUh3aXp8AAjWAwECIWQBAWAbIRn0keJlV5kANeV5EAPujADt-BNADF5QB-eUAClcTkAMm9ACY5MNAAPTQBVm21NYE0AAHTAEDIwAufUAG3jJtrOibzJOBKocH7sFp7B6cZ5n2gBtnwYsqdPE4VmvzpVw-ZuwEAB84DhhmwU8KCD2wfMZKckoY9cuO0+wCqGbgXlPGCv2A6DpmI59NHRXzKn-bpqqmbw6SBYDOBAEB3QBTV0APO1AAbnXlVNI7xoCgOBeDpPr+Lnula47hmu+biEOiAA)

- 方案 2：缺点是需要全局注册每个组件 Options 的类型，所幸可以用类型推断代替，但仍然需要一些样板代码。其优点是：将组件的类型绑定在了组件名上；即只要知道组件名，就能获得组件的完整类型提示。（虽然目前感觉这个优势不大）[代码](https://www.typescriptlang.org/zh/play?#code/LYewJgrgNgpgBAYxMADiAdjdAXAjHAbwCg5S4BLVEAJ2zgBU4BeRZNTHegTxRhLMppacACLNWVDtgDC1GAENsffqSToAznXTzg8FgHIkkrHgByOmPqIq4AeltxAWdqBJOUAjfoDe5QDD-gAHTAgZGAF40BZJUAKdQl2E25eOEBCK0B1dUBaOUALm0AYxTgAeRRscgx1OEAQt0BvH0Bo9RsYAA8hOjVNMIwTcREAOgQ5RRhpNjqcAFkYbHkwRXkACmIycbhtXQAaG3GQTOyNAC5COYnVTqlcESHVsY3D0jAYADN5aGwANXkoCBhVgAZZo-GAXxfDj-XSezhALATAOPxHkAsCqARh1AL8BgH+jQAMSulFjl8sUfnATNQuKthiBVvRGhksjkADzYHgwECnSYWAB8AEpmFS4CBPqQ3jSiG9rKBILBalIAExrcaCGh0RgsIzhTikmzC4RicVbEyyBRKazjapaCziQyKnB88y6Kw2P6ACUUXIyERobBqLQSNA1mq0lPilupev1Bv1RiiJV1sHzdv19ijxidzpcbncHnBnijvu82caHIArwMAFUqACuNALWm8LtuUKRUAbnqAH+1AF+KnkAp3KAbPlQr6pJF4ElkqUKiK4CSoh1jHqXYiWE0exo0qd3fJiaTybbXVS1WRypVWBoqrq6H3HSr2suR56RgcNlMYMyJgtc4fxmiMXAsTi8Zb1GPeBP97T6YybKz2ZzwNB4LWlU74LuFBUMIYq8hE0rNvOpwQOgCB2og-4Dm6fQDEMhJpKiZRKOgYC5LiACq6AANboCAADu6BIVSwwogAbteACSOAwNQ2hQEhW5oWk07jDSjHMaxtwcSh27oQygHjHI2AQKxcC0W+M6kHOrbQbB8EtOunaStgnH9ISpiYdhuEMI0Wl+gaMDqNRdHXmZUi6aOpg8WQfEmXZJgOfp4kolJMnoHJCmQa27bwEhQ4joS3SGVgxm4kxSiCext6efI6BcFSDIsDRRxReURl4Y08UsWxwkemh5DoKcLHpAyAD86RwKsmC0SxNhshyRBct+YFSlEgGynQ0gCgqXb+opdgOIAs4mAJ0OgBY-34wSACbWgAgmikgDz1oAVHIeHkOaunAyLjCFu05N08goOIElkH8gAvboApcYuD6y64BZwavGQv44DseyTBAwAAEatUc8aHH8gD+8oAFK5plm+aPaN+oWKsQ2mcufJIYeHVni2whHe5OAWbkLCaNQFUAOZwAAZHAREwFwE4cWd43KdjpLHRo+nRThuS49g+OZZeKIGXlMVc8u+NwPV9MoAA2qYAC6jWTDALXUG1jNY3QFUJecCDwIRJHkZRt6CocUv7qsROk7LqwwfrFGBZj85HUViWlahekYULnNwHrpEUVRF0ojLWoVVTNN01brOeWkQe6LLzksmrjss67omC1hwtwNzvPiNlhzOyVyUiVxt5efHcCJg7raayx2uhZanlXOINu++g3lHGGFxQNctz3KsVz27O6sUAJteZ5uRd6Wn+Vj6N2eXaQZtwKYKLHq6qwp2hTkouemKrzk68l05dJMAytEgOQYAD0pQ-V9Qo-cwAyn0WToCTd5TxnWcWFZxsbFg8h-VgGAVYf0QAgFgKlFet4D65i8lfCacBAD3yuWcGoJADtwYANeVAB90YAO39PAFkAChygAV+MAHtqgAeBUAKdGgBIc0ADHagAXU3zP4QAXJ5BWEKpOCSw4Akz6I-Z+pN34c2Ml-XQVlc4bA+tgRG48yqTwZAAHyXm1KRo0n7YBfm-Lyv8Jh-EAIDugBTV0AHnagAG5x8n0PyP0oBQDgPIXIzcDbWJFiovhr935lw5G8IAA)

注：上述 `Options` 类型是一种 `Object`，其内容是组件的 options 的“名称：值”的键值对，而非定义组件时传入的选项元信息。

**Comment by the1812** [comment]
我觉得方案 1 好些
v1 其实就是类似方案 2 的设计, 核心和功能太耦合了, 导致修改组件会有很多分散在核心的更改
